modyn/trainer_server/internal/trainer/pytorch_trainer.py

```diff
@@ -347,6 +349,82 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
+        sample_ids = batch[0]
+        if isinstance(sample_ids, torch.Tensor):
+            sample_ids = sample_ids.tolist()
+        elif isinstance(sample_ids, tuple):
+            sample_ids = list(sample_ids)
+
+        assert isinstance(sample_ids, list), "Cannot parse result from DataLoader"
+
+        if self._label_tranformer_function is None:
+            target = batch[2].to(self._device)
+        else:
+            target = self._label_tranformer_function(batch[2]).to(self._device)
+
+        data: Union[torch.Tensor, dict]
+        if isinstance(batch[1], torch.Tensor):
+            data = batch[1].to(self._device)
+        elif isinstance(batch[1], dict):
+            data: dict[str, torch.Tensor] = {}  # type: ignore[no-redef]
+            for name, tensor in batch[1].items():
+                data[name] = tensor.to(self._device)
+        else:
+            raise ValueError(
+                "The format of the data provided is not supported in modyn. "
+                "Please use either torch tensors or dict[str, torch.Tensor]"
+            )
+
+        return sample_ids, target, data
+
+    def sample_data(self) -> None:
+        """
+        Function to score every datapoint in the current dataset and sample a fraction of it
+        Used for downsampling strategies in sample_then_batch mode
+
+        """
+        assert self._downsampler is not None
+        assert self._sample_before_batch
+
+        # keys must be taken from the selector.
+        # This operation is needed only when we sample several times (otherwise the source is already the selector)
+        self._train_dataloader.dataset.switch_to_selector_key_source()
+
+        number_of_batches = 0
+
+        self._downsampler.setup_sample_then_batch()
+        for batch in self._train_dataloader:
+            number_of_batches += 1
+            sample_ids, target, data = self.prepare_data(batch)
+
+            with torch.autocast(self._device_type, enabled=self._amp):
+                # compute the scores and accumulate them
+                model_output = self._model.model(data)
+                self._downsampler.accumulate_sample_then_batch(model_output, target, sample_ids)
+
+        self._downsampler.end_sample_then_batch()
+
+        # to store all the selected (sample, weight).
+        file_size = self._num_dataloaders * self._batch_size  # should we add it to the pipeline?
+        local_dataset = LocalDatasetHandler(self.pipeline_id, self.trigger_id, self._num_dataloaders, file_size)
+
+        # each strategy can supply samples in different ways
+        samples_avilable = self._downsampler.samples_available()
+
+        while samples_avilable:
+            samples_list = self._downsampler.get_samples()
+            # store the selected samples (id and weight)
+            local_dataset.inform_samples(samples_list)
+            samples_avilable = self._downsampler.samples_available()
```

Comment:
sorry I do not understand the comment / what is happening here. could you elaborate?